File: 백준/Silver/9095. 1， 2， 3 더하기/1， 2， 3 더하기.java
import java.util.Scanner;

public class Main {
	public static void main(String[] args) {
		int[] numbers = scannerFunc();
		int[] dp = dpFunc();

		for (int num : numbers) {
			System.out.println(dp[num]);
		}
	}

	private static int[] scannerFunc() {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int[] numbers = new int[n];
		for (int i = 0; i < n; i++) {
			numbers[i] = sc.nextInt();
		}
		sc.close();
		return numbers;
	}

	private static int[] dpFunc() {
		int[] dp = new int[11];
		dp[1] = 1;
		dp[2] = 2;
		dp[3] = 4;

		for (int i = 4; i <= 10; i++) {
			dp[i] = dp[i - 3] + dp[i - 2] + dp[i - 1];
		}

		return dp;
	}
}

File: 백준/Silver/9095. 1， 2， 3 더하기/1， 2， 3 더하기.kt
import java.util.Scanner

fun main() {
    val numbers = scan()
    val dp = dp()
    for (i in numbers.indices) println(dp[numbers[i]])
}

fun scan(): IntArray = with(Scanner(System.`in`)) {
    val n = nextInt()
    val numbers = IntArray(n) {nextInt()}
    return numbers
}

fun dp() : IntArray {
    val dp = IntArray(11)
    dp[0] = 0
    dp[1] = 1
    dp[2] = 2
    dp[3] = 4
    for (i in 4 .. 10) dp[i] = dp[i - 3] + dp[i - 2] + dp[i - 1]
    return dp
}